Source/WTF/wtf/cocoa/TypeCastsCocoa.h

```diff
@@ -84,31 +84,88 @@ inline RetainPtr<id> bridge_id_cast(RetainPtr<CFTypeRef>&& object)
 #define checked_objc_cast checked_objc_castARC
 #endif
 
-template<typename T> inline T* checked_objc_cast(id object)
+template <typename ExpectedType>
+struct ObjCTypeCastTraits {
+public:
+    static bool isType(id _Nonnull object) { return [object isKindOfClass:[ExpectedType class]]; }
+
+    template <typename ArgType>
+    static bool isType(const ArgType * _Nonnull object) { return [object isKindOfClass:[ExpectedType class]]; }
+};
+
+#define SPECIALIZE_OBJC_TYPE_TRAITS_BEGIN(ClassName) \
+namespace WTF { \
+template <> \
+class ObjCTypeCastTraits<ClassName> { \
+public:
+
+#define SPECIALIZE_OBJC_TYPE_TRAITS_END() \
+}; \
+}
+
+template <typename ExpectedType, typename ArgType>
+inline bool is_objc(ArgType * _Nullable source)
+{
+    return source && ObjCTypeCastTraits<ExpectedType>::isType(source);
+}
+
+template<typename T> inline T *_Nullable checked_objc_cast(id _Nullable object)
+{
+    if (!object)
+        return nullptr;
+
+    RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(is_objc<T>(object));
+
+    return reinterpret_cast<T*>(object);
+}
+
+template<typename T, typename U> inline T *_Nullable checked_objc_cast(U *_Nullable object)
 {
     if (!object)
         return nullptr;
 
-    RELEASE_ASSERT_WITH_SECURITY_IMPLICATION([object isKindOfClass:[T class]]);
+    RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(is_objc<T>(object));
 
     return reinterpret_cast<T*>(object);
 }
 
 // Use dynamic_objc_cast<> instead of checked_objc_cast<> when actively checking NS types,
-// similar to dynamic_cast<> in C++. Be sure to include a nil check.
+// similar to dynamic_cast<> in C++.
 
 // See RetainPtr.h for: template<typename T> T* dynamic_objc_cast(id object).
 
 template<typename T, typename U> RetainPtr<T> dynamic_objc_cast(RetainPtr<U>&& object)
 {
-    if (![object isKindOfClass:[T class]])
+    if (!is_objc<T>(object.get()))
         return nullptr;
-
     return WTFMove(object);
 }
 
+template<typename T, typename U> RetainPtr<T> dynamic_objc_cast(const RetainPtr<U>& object)
```

Comment:
You should fix up the `dynamic_objc_cast()` overload in `RetainPtr.h` as well.  Or is there a different plan for that?

(The reason it hasn't moved to `TypeCastsCocoa.h` is that there were uses of it in WebKit and the internal project that required updating `#include` headers to add `#include <wtf/cocoa/TypeCastsCocoa.h>`.)